peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+
+   # NOTE: A more straightforward implementation would use isinstance()
+   def ensure_int(value: object) -> None:
+       value_type: TypeExpr = type(value)  # assigns a type (a subtype of TypeExpr)
+       assert value_type == int
+
+
+TypeExpr Values
+---------------
+
+A variable of type ``TypeExpr[T]`` where ``T`` is a type, can hold any
+**type expression object** - the result of evaluating a `type expression`_
+at runtime - which is a subtype of ``T``.
+
+Incomplete expressions like a bare ``Optional`` or ``Union`` which do
+not spell a type are not ``TypeExpr`` values.
+
+``TypeExpr[...]`` is itself a ``TypeExpr`` value:
+
+::
+
+   OPTIONAL_INT_TYPE: TypeExpr = TypeExpr[Optional[int]]  # OK
+   assert isassignable(Optional[int], OPTIONAL_INT_TYPE)
+
+
+Stringified TypeExprs
+'''''''''''''''''''''
+
+A ``TypeExpr`` value may be a string literal that spells a forward
+reference:
+
+::
+
+   IntTreeRef: TypeExpr = 'IntTree'  # OK
+
+However the string itself must spell a valid type to be considered a
+``TypeExpr``:
+
+::
+
+   BadUnion1: TypeExpr = Union    # ERROR: does not spell a type
+   BadUnion2: TypeExpr = 'Union'  # ERROR: does not spell a type
+
+
+Literal[] TypeExprs
+'''''''''''''''''''
+
+To simplify static type checking, a ``Literal[...]`` value is *not*
+considered assignable to a ``TypeExpr`` variable even if all of its members
+spell valid types:
+
+::
+
+   STRS_TYPE_NAME: Literal['str', 'list[str]'] = 'str'
+   STRS_TYPE: TypeExpr = STRS_TYPE_NAME  # ERROR: Literal[] value is not a TypeExpr
+
+However ``Literal[...]`` itself is still a ``TypeExpr``:
+
+::
+
+   DIRECTION_TYPE: TypeExpr = Literal['left', 'right']  # OK
+
+
+.. _non_universal_typeexpr:
+
+Non-Universal TypeExprs
+'''''''''''''''''''''''
+
+``TypeExpr[]`` values include *all* type expressions including some
+which are not valid in all annotation contexts. In particular:
+
+-  ``Self`` (valid only in some contexts)
+-  ``TypeGuard[...]`` (valid only in some contexts)
+-  ``TypeIs[...]`` (valid only in some contexts)
+
+``Self`` can be used as a function parameter type, a return type, or a
+variable type, but only within a class definition:
+
+::
+
+   class ReturnsSelf:
+       def __call__(self: Self) -> Self:  # OK
+           return self
+
+   class HoldsSelf:
+       me: Self  # OK
+
+   def returns_self(me: Self) -> Self: ...  # ERROR: Self meaningless outside class
+
+``TypeGuard[]`` and ``TypeIs[]`` can only be used as a return type of a
+function. They cannot be used as a variable type or a parameter type:
+
+::
+
+   def is_positive_int(value: object) -> TypeGuard[int]: ...  # OK
+
+   def nonsense(value: TypeGuard[int]): ...  # ERROR: TypeGuard[] not meaningful here
+
+   exotic_bool: TypeGuard[int]  # ERROR: TypeGuard[] not meaningful here
+
+Non-universal type expressions may be treated as equivalent to ``Any``
+for inference purposes:
+
+::
+
+   typx: TypeExpr = TypeIs[str]
+   assert_type(typx, TypeExpr[Any])  # acceptable
```

Comment:
This seems like an odd special case. Why is it necessary?

Reply:
The whole concept of allowing non-universal type expressions to parameterize/constrain a type variable doesn't make a lot of sense to me, so I'm attempting to give type checkers leeway to immediately convert non-universal type expressions to `Any`, both immediately (in this section) and when in combination with type variables (in §"Interactions with type variables").

Do you think it would be better if I simply don't mention non-universal type expressions at all and leave their behavior unspecified, by omission?

Reply:
I think it's fine to leave this undefined.

Reply:
I've now edited out any specification of _behavior_ of non-universal type expressions, but still left in the _definition_ of a  "non-universal type expression" since I expect it may be a useful concept to be able to name.